### PR TITLE
Added monkey-patch links to html/fetch

### LINKED ISSUE #24: Subresource request prioritization of <iframe> and <script>
Both <script> and <iframe> DOM elements can trigger a lot of resource requests. It's true we don't have a real *processing model* in this document, but we should still probably define here what effects the `importance` value given to a top-level <script> or <iframe> will have on their requested subresources. *Note: this is different than the dependency graph request prioritization of a module script which is outside of the scope of this discussion*.

Here are a few questions:

## Should subresources be given the same `importance` value as their parent?
In other words, should every request made in the below example all have `importance = low`

```
// index.html
<iframe importance=low src=doc.html></iframe>
<script importance=low src=main.js></script>
```
```
// doc.html
<img src=https://image.here alt=image>
```
```
// main.js
fetch("https://domfarolino.com")
```
This definitely seems logical to me to get the most benefit out of the Priority Hints effort.

## What if subresources have conflicting `importance` values?
In other words, what are the resolved `importance` values of all of the below requests:

```
// index.html
<iframe importance=low src=doc.html></iframe>
<script importance=low src=main.js></script>
```
```
// doc.html
<img importance=high src=https://image.here alt=image>
```
```
// main.js
fetch("https://domfarolino.com", {importance: high})
```

I'm not sure what the best call to make here is. In my opinion, the two best options are:
 - Leave it up to implementations
 - Ignore overridden subresource `importance` values, they should just assume that of their parent

## What about a script that adds an element to the DOM with a higher importance than the script itself?

```
<script importance=low src=main.js>
```
```
// main.js
let link = document.createElement('link');
link.importance = 'high';
link.......
.....
document.head.appendChild(link);
```
I'm not even sure if we could bully-down the added element's priority based on *who* added it to the DOM, but I don't think we should do that in this case. IMO I think it would make sense if the `<link>`'s ultimate `importance` was `high` here.

...more example? Thoughts?

I assume it would be good to specify this in the document so people can see how far-reaching and influential an `importance` value is on these types of elements.

### LINKED ISSUE #28: Update EXAMPLES.md and EXPLAINER.md to reflect spec
The Examples file mentions `importance=critical` and `importance=medium` but the proposed spec (https://wicg.github.io/priority-hints/) only mentions setting importance with either high or low.

Should `EXAMPLES.md` be updated to reflect the latest spec?

### LINKED ISSUE #29: Document "safety" of applying priority hints
After chatting with @domfarolino today, an item we noted would be useful for the explainer is the safety of using priority hints. 

* Hints should be safe in that hints specified by a site shouldn't degrade performance both (1) in a browser that supports `importance` and (2) in a browser that does not. 
* UAs should feel free to apply other heuristics in addition to hints to decide on how they load content. e.g in Chrome, with hero images Chrome may automatically detect the image is important if it gets to layout early enough but all browsers would benefit from knowing which images you want loaded "earlier" and which are safe to load "later", regardless of how they use this information.
* This same logic applies for scripts. Browsers may have different heuristics that work differently across browsers for scripts (e.g with async, defer etc) but a site could fix any mis-guesses across the board by being explicit using hints. Applying hints shouldn't negatively impact the overall experience.

### LINKED ISSUE #41: Bring back examples into the explainer
While answering questions on the [TAG review](https://github.com/w3ctag/design-reviews/issues/241#issuecomment-474800105), it seemed like the usage examples have moved from the explainer to the spec itself. While it's important to have them in the spec, it seems equally important to have them as part of the explainer. We should probably copy them back.

### PR DESCRIPTION
**Pull request by pmeenan**
This is a first attempt at adding spec references to HTML and fetch for the relevant spec changes.  Verified with bikeshed that all of the links are connected properly and it compiles without warning.

Also cleaned up the references to image prioritization and preload to better map to how Chrome behaves and removed the section about the importance cascading to iframe subresources. Currently only focused on the main resource fetch scheduling.

fixes #41, fixes #29, fixes #28, closes #24

### CODE REVIEW

**Comment by pmeenan** [comment]
@yoavweiss could you PTAL and see if this is on the correct path for updating the spec with monkey-patch references?

@domfarolino and @addyosmani would you mind taking a look at the spec changes to make sure they make sense?

File: index.bs [hunk 1]
```diff
@@ -5,8 +5,9 @@ Level: 1
 Status: CG-Draft
 Group: WICG
 URL: https://wicg.github.io/priority-hints
-Editor: Addy Osmani, Google Inc., addyosmani@gmail.com, https://blog.yoav.ws
+Editor: Addy Osmani, Google Inc., addyosmani@gmail.com, https://addyosmani.com/
```
Comment (yoavweiss) [inline index.bs:9]: So that's why I was getting all that traffic :D

File: index.bs [hunk 2]
```diff
@@ -66,9 +72,9 @@ spec:html; type:dfn; for:/; text:missing value default
       up their discovery using <code>&lt;link rel=preload&gt;</code>([[PRELOAD]]). Browsers mostly determine a request's priority based on the
       request's <a>destination</a>, and location in the containing document if applicable.</p>
 
-    <p>This document details use cases and a markup that will provide developers control to indicate a resource's
-      relative importance to the browser, enabling the browser to act on those indications to influence the request's overall priority
-      in ways described in the <a href="#effects-of-priority-hints">Effects of Priority Hints</a> section.
+    <p>This document details use cases and modifications to [[fetch#fetch-method|fetch]] and <code>HTML</code> markup that will provide developers
```
Comment (yoavweiss) [inline index.bs]: Here you want to refer to the Fetch and HTML specs, not specific algorithms. You can do that with `[[FETCH]]` and `[[HTML]]`.

File: index.bs [hunk 3]
```diff
@@ -113,6 +119,68 @@ spec:html; type:dfn; for:/; text:missing value default
     <p>This is how we conceptually think about different resource types under the hood in browsers today. It may translate well
       to user-space where different types of content share similar properties.</p>
 
+    <h2 id="proposal">Proposal</h2>
+    <h3 id="fetch">Fetch</h3>
+    <ul>
+      <li>Add an <dfn id="enum-fetch-request-importance" export>Importance enum</dfn> as an <a>enumerated attribute</a> of the {{Request|Request}} class.
```
Comment (yoavweiss) [inline index.bs]: You probably want to have a "definitions" section that'd contain the `enum Importance` dfn, and then turn this section into "Fetch integration".

Then the way we typically extend objects defined in other specs with something like: 'A [=Request=] has an |importance|, an [=Importance=], initially set to "auto".'

File: index.bs [hunk 4]
```diff
@@ -113,6 +119,68 @@ spec:html; type:dfn; for:/; text:missing value default
...
+        <pre class=idl>
+        enum Importance { "high", "low", "auto" };
+        </pre>
+      </li>
+
+      <li>Add a readonly <code>importance</code> attribute to the {{Request|Request}} interface of the type {Importance}.</li>
+      <li>Add a readonly <code>importance</code> attribute to the {{Request|RequestInit}} dictionary of the type {Importance}.</li>
+    </ul>
+    <p>The attribute's <i data-x="missing value default">missing value default</i> and <i
```
Comment (yoavweiss) [inline index.bs]: I think we want to define that as part of the algorithms that set the attribute, both initially and as part of error handling.

**Review by yoavweiss — COMMENTED** [review]
A couple of initial comments. I'll continue to review later

File: index.bs [hunk 5]
```diff
@@ -113,6 +119,68 @@ spec:html; type:dfn; for:/; text:missing value default
...
+  data-x="invalid value default">invalid value default</i> are both the <code>auto</code> importance.</p>
+    <p>The Request importance can be interpreted by the underlying browser engine to influence the prioritization of the fetch.</p>
+
+    <h3 id="html">HTML</h3>
```
Comment (yoavweiss) [inline index.bs]: Maybe on top of this table, I think we want to actually define what the attribute mean in terms of their processing model:
* You want to define an algorithm in HTML that take and element as input and turns its `importance` attribute value (DOMString) into an importance enum value.
* Then in [update the image data](https://html.spec.whatwg.org/multipage/images.html#update-the-image-data) step 17, you want to add a step that defines the request's importance attribute to the result of that algorithm, called with the element as input.
* in [link type "preload"](https://html.spec.whatwg.org/multipage/links.html#link-type-preload) you want to add a similar step
* Same for [link type "stylesheet"](https://html.spec.whatwg.org/multipage/links.html#link-type-stylesheet) where you probably want to add to its "linked resource fetch setup steps" a step that adds the importance to the request.
* Scripts would be a bit trickier, where as part of [prepare a script](https://html.spec.whatwg.org/multipage/scripting.html#prepare-a-script) you'd want to create the importance enum value and then pass it to [Fetch a classic script](https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-classic-script) and [Fetch an external module script graph](https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-module-script-tree) where you'd set the value on the requests these algorithms create.

**Comment by pmeenan** [comment]
None of the processing models have been added yet (going to start working on those now) but I think the attribute and interface extensions are now correctly monkey-patched back to the fetch and HTML specs.

File: index.bs [hunk 6]
```diff
@@ -5,16 +5,26 @@ Level: 1
 Status: CG-Draft
 Group: WICG
 URL: https://wicg.github.io/priority-hints
-Editor: Addy Osmani, Google Inc., addyosmani@gmail.com, https://blog.yoav.ws
+Editor: Addy Osmani, Google Inc., addyosmani@gmail.com, https://addyosmani.com/
 Editor: Yoav Weiss 58673, Google Inc., yoav@yoav.ws, https://blog.yoav.ws
+Editor: Patrick Meenan, Google Inc., patmeenan@gmail.com, https://twitter.com/patmeenan
```
Comment (yoavweiss) [inline index.bs]: I think you should bump yourself to the top of the list, as the current primary editor

**Review by yoavweiss — COMMENTED** [review]
This LGTM as an improvement on the current spec, even if many things are still missing. Do you want to land this and iterate the processing model over a separate PR?

**Comment by yoavweiss** [comment]
For the [build error](https://app.travis-ci.com/github/WICG/priority-hints/builds/243144613) it seems like we need to update the python version in travis.yml

File: index.bs (hunk 6 above)
Comment (pmeenan) [reply index.bs]: Done

**Comment by pmeenan** [comment]
@yoavweiss Can you PTAL at the Travis modifications? It looks like travis added native deploy support to GH pages since this was originally done so it seems a fair bit cleaner.

I got rid of the deploy script and updated it to match the travis script used for the [conversion-measurement-api](https://github.com/WICG/conversion-measurement-api/blob/main/.travis.yml) spec (using the correct decryption key for this project) and configured it to treat warnings as errors since the spec currently compiles cleanly.

**Review by yoavweiss — APPROVED** [review]
LGTM - including the travis parts. I think it'd be better to complete the processing model TODOs in a separate PR.

**Pull request merged**